src/Microsoft.TestPlatform.CommunicationUtilities/DataCollectionRequestHandler.cs

```diff
@@ -372,17 +393,27 @@ private void HandleAfterTestRunEnd(Message message)
             }
 
             var attachmentsets = this.dataCollectionManager.SessionEnded(isCancelled);
+            var afterTestRunEndResult = new AfterTestRunEndResult(attachmentsets, this.requestData.MetricsCollection.Metrics);
 
             // Dispose all datacollectors before sending attachments to vstest.console process.
             // As datacollector process exits itself on parent process(vstest.console) exits.
             this.dataCollectionManager?.Dispose();
 
-            this.communicationManager.SendMessage(MessageType.AfterTestRunEndResult, attachmentsets);
+            this.communicationManager.SendMessage(MessageType.AfterTestRunEndResult, afterTestRunEndResult);
             EqtTrace.Info("DataCollectionRequestHandler.ProcessRequests : Session End message received from server. Closing the connection.");
 
             this.Close();
 
             EqtTrace.Info("DataCollectionRequestHandler.ProcessRequests : DataCollection completed");
         }
+
+        private void UpdateRequestData(bool isTelemetryOptedIn)
+        {
+            if (isTelemetryOptedIn && !this.requestData.IsTelemetryOptedIn)
```

Comment:
nit: This looks like it would be better off sitting inside RequestData class. You could also make it more generic by updating it to be `if (isTelemetryOptedIn != this.requestData.IsTelemetryOptedIn` so the intent of this function is clear. 
I would also then change the assignment below.

Reply:
Changed logic to be more general